asset-transfer-private-data/chaincode-go/private_asset_transfer.go

```diff
@@ -0,0 +1,501 @@
+
+	transMap, err := ctx.GetStub().GetTransient()
+	if err != nil {
+		return fmt.Errorf("Error getting transient: " + err.Error())
+	}
+
+	// Asset properties are private, therefore they get passed in transient field
+	transientDeleteJSON, ok := transMap["agree_delete"]
+	if !ok {
+		return fmt.Errorf("asset to delete not found in the transient map")
+	}
+
+	type assetDelete struct {
+		ID string `json:"asset_id"`
+	}
+
+	var assetDeleteInput assetDelete
+	err = json.Unmarshal(transientDeleteJSON, &assetDeleteInput)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal JSON: %s", err.Error())
+	}
+
+	if len(assetDeleteInput.ID) == 0 {
+		return fmt.Errorf("ID field must be a non-empty string")
+	}
+
+	// delete private details of agreement
+
+	orgCollection, err := getCollectionName(ctx, true) // get proposers collection
+
+	err = ctx.GetStub().DelPrivateData(orgCollection, assetDeleteInput.ID) // delete the asset
+	if err != nil {
+		return err
+	}
+
+	// delete transfer agreement record
+
+	tranferAgreeKey, err := ctx.GetStub().CreateCompositeKey(transferAgreementObjectType, []string{assetDeleteInput.ID}) // create composite key
+	if err != nil {
+		return fmt.Errorf("failed to create composite key: %s", err.Error())
+	}
+
+	err = ctx.GetStub().DelState(tranferAgreeKey) // remove agreement from state
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+// getCollectionName is an internal helper function to get collection of submitting client identity.
+// The collection name can optionally be verified against the peer org ID, to ensure that a
+// client from another org doesn't attempt to read or write private data from this peer.
+func getCollectionName(ctx contractapi.TransactionContextInterface, verifyOrg bool) (string, error) {
+
+	// Get the MSP ID of submitting client identity
+	clientMSPID, err := ctx.GetClientIdentity().GetMSPID()
+	if err != nil {
+		return "", fmt.Errorf("failed to get verified OrgID: %s", err.Error())
+	}
+
+	// verify that the client is submitting request to peer in their organization
+	if verifyOrg {
+		err = verifyClientOrgMatchesPeerOrg(clientMSPID)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// create the collection name
+	orgCollection := clientMSPID + "PrivateCollection"
+
+	return orgCollection, nil
+}
+
+// verifyClientOrgMatchesPeerOrg is an internal function used verify client org id and matches peer org id.
+func verifyClientOrgMatchesPeerOrg(clientMSPID string) error {
+	peerMSPID, err := shim.GetMSPID()
+	if err != nil {
+		return fmt.Errorf("failed getting peer's orgID: %s", err.Error())
+	}
+
+	if clientMSPID != peerMSPID {
+		return fmt.Errorf("client from org %s is not authorized to read or write private data from an org %s peer", clientMSPID, peerMSPID)
+	}
+
+	return nil
+}
+
+func main() {
+
+	chaincode, err := contractapi.NewChaincode(new(SmartContract))
+
+	if err != nil {
+		fmt.Printf("Error creating private mables chaincode: %s", err.Error())
```

Comment:
Panicf without a return is better

import "log"
log.Panicf("Error creating asset-transfer-private-data chaincode: %v", err)

Reply:
done.